### PR TITLE
reorganize channel handshake handler

### PR DESCRIPTION
**Pull request by colin-axner**
## Description

ref: #640 

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [x] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/ibc-go/blob/master/CONTRIBUTING.md#pr-targeting))
- [x] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [x] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules/structure.md).
- [ ] Wrote unit and integration [tests](https://github.com/cosmos/ibc-go/blob/master/CONTRIBUTING.md#testing)
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`)
- [x] Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [x] Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [x] Re-reviewed `Files changed` in the Github PR explorer
- [ ] Review `Codecov Report` in the comment section below once CI passes

### CODE REVIEW

File: modules/core/04-channel/keeper/handshake.go [hunk 1]
```diff
@@ -82,7 +100,12 @@ func (k Keeper) ChanOpenInit(
 		),
 	})
 
-	return channelID, capKey, nil
+	ctx.EventManager().EmitEvents(sdk.Events{
```
Comment (colin-axner) [inline modules/core/04-channel/keeper/handshake.go:103]: someone can do a followup pr to add a function `EmitOpenInitEvents` to the `04-channel/keeper/events.go` file for each handshake step

File: modules/apps/27-interchain-accounts/controller/keeper/handshake.go [hunk 2]
```diff
@@ -104,8 +104,9 @@ func (k Keeper) OnChanCloseConfirm(
 
 // validateControllerPortParams asserts the provided connection sequence and counterparty connection sequence
 // match that of the associated connection stored in state
-func (k Keeper) validateControllerPortParams(ctx sdk.Context, channelID, portID string, connectionSeq, counterpartyConnectionSeq uint64) error {
-	connectionID, connection, err := k.channelKeeper.GetChannelConnection(ctx, portID, channelID)
+func (k Keeper) validateControllerPortParams(ctx sdk.Context, connectionHops []string, connectionSeq, counterpartyConnectionSeq uint64) error {
```
Comment (colin-axner) [inline modules/apps/27-interchain-accounts/controller/keeper/handshake.go:107]: applications must not rely on channel state being set before the application callback

File: modules/core/04-channel/keeper/handshake.go [hunk 3]
```diff
@@ -373,6 +440,21 @@ func (k Keeper) ChanOpenConfirm(
...
+// WriteOpenConfirmChannel writes an updated channel state for the successful OpenConfirm handshake step.
+// An event is emitted for the handshake step.
+func (k Keeper) WriteOpenConfirmChannel(
+	ctx sdk.Context,
+	portID,
+	channelID string,
+) {
+	channel, found := k.GetChannel(ctx, portID, channelID)
+	if !found {
+		panic(fmt.Sprintf("could not find existing channel when updating channel state in successful ChanOpenAck step, channelID: %s, portID: %s", channelID, portID))
```
Comment (colin-axner) [inline modules/core/04-channel/keeper/handshake.go:455]:
```suggestion
		panic(fmt.Sprintf("could not find existing channel when updating channel state in successful ChanOpenConfirm step, channelID: %s, portID: %s", channelID, portID))
```

**Review by AdityaSripal — APPROVED** [review]

**Review by seantking — APPROVED** [review]

File: modules/apps/27-interchain-accounts/controller/keeper/handshake.go (hunk 2 above)
Comment (seantking) [reply modules/apps/27-interchain-accounts/controller/keeper/handshake.go:107]: should we document this somewhere? :thinking:

Comment (colin-axner) [reply modules/apps/27-interchain-accounts/controller/keeper/handshake.go:107]: I'll update the docs when I do the OnChanOpenTry change

Comment (seantking) [reply modules/apps/27-interchain-accounts/controller/keeper/handshake.go:107]: Cool. I was just thinking out loud :)

**Pull request merged**